### ISSUE TITLE
IDE-extension Big Bug -> Namespaces are not parsed correctly

### ISSUE THREAD

**Issue by NilsJacobsen** [inlang/inlang#1012]
## Problem

Namespaces are not parsed like i18next wants it to do. Which brakes all projects with namespace. cal.com/ open-assistent ...
https://www.loom.com/share/d5dce1740327498cb839a565589ac215?sid=eeb0141e-0faa-472e-85eb-b73cc904fc7d

https://www.i18next.com/principles/namespaces

@inlang/ide-extension

**Comment by samuelstroschein** [comment]
Oh god, this is hell. 

A few thoughts: 

- adding namespaces to `core` seems unavoidable. we know it must come, the longer we prolong it, the more hacky workarounds with risk for bugs we'll introduce. 
- plugins should define how messages are read. Hardcoding stuff in the ide extension will lead to maintenance overhead and constant "edge case support bugfixing". 
- a message reference match should be able to return a namespace https://github.com/inlang/inlang/blob/main/source-code/core/src/config/ideExtension/schema.ts#L53

**Comment by felixhaeberle** [comment]
> adding namespaces to core seems unavoidable

this is the only way to handle it correctly long-term.

> Hardcoding stuff in the ide extension will lead to maintenance overhead and constant "edge case support bugfixing".

💪

> a message reference match should be able to return a namespace https://github.com/inlang/inlang/blob/main/source-code/core/src/config/ideExtension/schema.ts#L53

✅

**Comment by samuelstroschein** [comment]
Summary from the call: 

- the message reference parser in the plugin must be adjusted in any case to detect namespaces
- quick-fix: adjust the parser in the plugin and try to solve the issue without proper namespace support in core

**Comment by NilsJacobsen** [comment]
So tasks for now:
- add the namespace to the key with an `:` instead of an `.` + add test
- detect file-based formatting of namespaces (with `t( key, { ns: namespace })` or`t(namespace:key)`)
- Adjust parser to read either prefix or option syntax to correctly recognize translations
- Adjust the serializer to extract messages according to the common syntax

**Comment by felixhaeberle** [comment]
@NilsJacobsen could it be that you created the same issue, #1012 & #1014 twice?

**Comment by NilsJacobsen** [comment]
I don't think so. this one is about the parsing of the plugins (also in current version) and the other one is about the new plugin and why the ide extension brakes there

**Comment by felixhaeberle** [comment]
as @samuelstroschein summarized from the call:

> Summary from the call:
> - the message reference parser in the plugin must be adjusted in any case to detect namespaces
> - quick-fix: adjust the parser in the plugin and try to solve the issue without proper namespace support in core

any further steps & improvements can be evaluated until we reach namespace support in the parser 👍

**Comment by NilsJacobsen** [comment]
Update: Yesterday I was pretty successful with parsing the namespace. Gonna finish that on Monday.

**Comment by NilsJacobsen** [comment]
Update: 
 - We can parse both colon and object syntax from files
 - In both cases, the messageId gets prefixed with a colon when extracting 
 - For now, I would not add the detection until people complain
 
Question: How are we gonna deal with default namespaces?
-> We could check in the config file if the pathPattern is a string or an object. We can prefix the keys with the first prefix in the list when it is an object. 
What do you think about that quick fix? Of course, we need to document that properly.

**Comment by NilsJacobsen** [comment]
Is it possible to access the config in the ideExtension functions?

**Comment by NilsJacobsen** [comment]
Sorry, I was a bit unclear. I mean, if you could access the PluginSettings inside the ide-extension parser inside the plugin. Cause for the read and write functions, the settings get passed directly via args. The general type of the ideExtensionConfig doesn't allow that, right?

**Comment by NilsJacobsen** [comment]
Okay, I feel like we do not want the parser to deal with the config stuff. It should just stupidly parse, right? So another way would be to check in the writeFunction, if there is a key with no namespace when the pathPattern requires it. And then add the first namespace.

**Comment by NilsJacobsen** [comment]
Okay, it is implemented as follows:
- Write function checks if there needs to be a namespace.
- If not provided it adds the first namespace in the list

That should fix the problem for now.

**Comment by NilsJacobsen** [comment]
Wait, that helps to add a namespace when people are extracting strings with a prefix, but it doesn't help the ide extension to access the correct file when there is no prefix in the key at all. 

I do not want to ship that detection into the ide-extension, so we must access the pathPattern data in the parser. 

@felixhaeberle @openscript can we pass the settings in the parser like this:

import type { InlangConfig } from "@inlang/core/config"
import { parse } from "./messageReferenceMatchers.js"

```ts
export const ideExtensionConfig: InlangConfig["ideExtension"] = {
	messageReferenceMatchers: [
		async (args) => {
			return parse(args.documentText, settings) // <- adding settings here to access it
		},
	],
	extractMessageOptions: [
		{
			callback: (messageId) => `{t("${messageId}")}`,
		},
	],
	documentSelectors: [
		{
			language: "javascript",
		},
		{
			language: "typescript",
		},
		{
			language: "svelte",
		},
	],
}
```

What would that mean for the ide-extension?

**Comment by openscript** [comment]
How is `settings` provided? Where does it come from?

**Comment by NilsJacobsen** [comment]
That was just a pseudo-code example (more or less). As far as I understood, these functions are used by the ide-extension, so we could use the `state().config` like u supposed in the ide-extension and pass the pluginSettings into the parse function.

**Comment by openscript** [comment]
> That was just a pseudo-code example (more or less). As far as I understood, these functions are used by the ide-extension, so we could use the `state().config` like u supposed in the ide-extension and pass the pluginSettings into the parse function.

Yes, that should work.

**Comment by samuelstroschein** [comment]
@NilsJacobsen super confusing updates. I don't know what exactly you are asking for. 

What I understand is that you want to access the plugin settings in the parser of the plugin, you can do that. Just pass the settings from the `createPlugin` function to the ide extension settings. 

https://github.com/inlang/inlang/blob/907b035687d0dd8d5d5a3a0cfce65d55a6c06983/source-code/plugins/i18next/src/plugin.ts#L92-L95

> @NilsJacobsen  As far as I understood, these functions are used by the ide-extension, so we could use the state().config like u supposed in the ide-extension and pass the pluginSettings into the parse function.

No, you can't and shouldn't. 

You are confusing app state with plugin logic. Plugins have no access to whatever logic applications come up with. `state().config` is APPLICATION logic that you will never be able to access from a plugin. 

Can you explain where the confusion is coming from?

**Comment by NilsJacobsen** [comment]
I guess I was thrown into cold water by working on the ide-extension and understanding every aspect of it. But I guess it gets clearer. 

But jap you totally got what I wanted to achieve. So I gonna do it that way.

**Comment by NilsJacobsen** [comment]
Now it works except they use `next-i18next` hook `const t = useTranslation("namespace")`. That way, they can set the namespace scoped to the t functions in that file. 

I would add this parsing function if somebody asks for it. Meanwhile, I'm gonna close this.
